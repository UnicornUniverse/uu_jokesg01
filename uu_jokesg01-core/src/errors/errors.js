// devkit-pure-exports
export * from "./get-error-lsi.js";
export * from "./get-error-status.js";
export * from "./property-error.js";
export * from "./uu-jokes-error.js";
export * from "./http-status.js";
