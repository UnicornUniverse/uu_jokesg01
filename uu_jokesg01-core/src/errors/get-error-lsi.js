//@@viewOn:imports
import HttpStatus from "./http-status";
import getErrorStatus from "./get-error-status";
import LsiData from "./errors-lsi";
//@@viewOff:imports

function getErrorLsi(errorData, customErrorLsi) {
  let lsi = getErrorMessage(errorData, customErrorLsi);

  if (!lsi) {
    const errorStatus = getErrorStatus(errorData);
    lsi = getErrorMessageByStatus(errorStatus, customErrorLsi);
  }

  return lsi;
}

function getErrorMessageByStatus(errorStatus, customErrorLsi) {
  let lsi;

  switch (errorStatus) {
    case HttpStatus.BaseNetworkError:
      lsi = customErrorLsi.baseNetworkError || LsiData.baseNetworkError;
      break;
    case HttpStatus.BadRequest:
      lsi = customErrorLsi.badRequest || LsiData.badRequest;
      break;
    case HttpStatus.Unauthorized:
      lsi = customErrorLsi.unauthorized || LsiData.unauthorized;
      break;
    case HttpStatus.Forbidden:
      lsi = customErrorLsi.forbidden || LsiData.forbidden;
      break;
    case HttpStatus.NotFound:
      lsi = customErrorLsi.notFound || LsiData.notFound;
      break;
    case HttpStatus.InternalServerError:
      lsi = customErrorLsi.internal || LsiData.internal;
      break;
    case HttpStatus.ServiceUnavailable:
      lsi = customErrorLsi.serviceUnavailable || LsiData.serviceUnavailable;
      break;
    case HttpStatus.GatewayTimeout:
      lsi = customErrorLsi.requestTimeout || LsiData.requestTimeout;
      break;
    default:
      lsi = customErrorLsi.defaultError || LsiData.defaultError;
  }

  return lsi;
}

function getErrorMessage(errorData, customErrorLsi) {
  const code = errorData?.error?.code || errorData.code;
  return customErrorLsi[code] || LsiData[code] || errorData?.error?.message;
}

//@@viewOn:exports
export { getErrorLsi };
export default getErrorLsi;
//@@viewOff:exports
