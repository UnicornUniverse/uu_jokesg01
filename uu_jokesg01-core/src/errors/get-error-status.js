export function getErrorStatus(errorData) {
  // note: there were cases when errorData without reparsing
  // were not behaving like an object
  // const errorData = JSON.parse(JSON.stringify(props.errorData));
  let status = errorData?.status;

  if (status === null || status === undefined) {
    return errorData?.error?.status;
  } else {
    return status;
  }
}

export default getErrorStatus;
