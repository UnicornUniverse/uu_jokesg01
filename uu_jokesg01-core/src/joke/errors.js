import Config from "./config/config";
import PropertyError from "../errors/property-error";

const ERROR_PREFIX = Config.TAG.toLowerCase().replaceAll(".", "-") + "provider/";

export class NoOidError extends PropertyError {
  static code = ERROR_PREFIX + "no-oid";

  constructor(cause) {
    super(NoOidError.code, "The required property oid is not defined!", cause);
  }
}

export default { NoOidError };
