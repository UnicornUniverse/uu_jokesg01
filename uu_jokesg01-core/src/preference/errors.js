import Config from "./config/config";
import PropertyError from "../errors/property-error";

const ERROR_PREFIX = Config.TAG.toLowerCase().replaceAll(".", "-") + "provider/";

export class NoUu5IdError extends PropertyError {
  static code = ERROR_PREFIX + "no-uu5-id";

  constructor(cause) {
    super(NoUu5IdError.code, "The required property Component identifier (uu5Id) is missing!", cause);
  }
}

export default { NoUu5IdError };
