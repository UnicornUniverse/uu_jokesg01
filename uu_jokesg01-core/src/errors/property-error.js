import { UuJokesError } from "../errors/uu-jokes-error";

export class PropertyError extends UuJokesError {
  constructor(code, message, cause) {
    super(code, message, cause);
  }
}

export default PropertyError;
