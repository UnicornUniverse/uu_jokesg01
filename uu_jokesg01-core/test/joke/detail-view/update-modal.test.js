import { Client } from "uu_appg01";
import { Test, VisualComponent, omitConsoleLogs } from "uu5g05-test";
import UpdateModal from "../../../src/joke/detail-view/update-modal";
import { UuJokesError } from "../../../src/errors/errors";
import queries from "../../queries/queries";
import enLsi from "../../../src/lsi/en.json";

const { screen } = Test;
const lsi = enLsi[UpdateModal.uu5Tag];

const categoryList = [
  { id: "1", name: "Category 1" },
  { id: "2", name: "Category 2" },
  { id: "3", name: "Category 3" },
];

function getDefaultProps() {
  const jokeDataObject = {
    state: "ready",
    data: {
      id: "123",
      name: "Test joke",
      visibility: false,
      text: "Best joke ever",
      image: "image-1",
      imageFile: new File([], "image-1.jpg", { type: "image/jpeg" }),
      ratingCount: 5,
      averageRating: 2.5,
      categoryIdList: ["1", "2"],
      categoryList: [
        { id: "1", name: "Category 1" },
        { id: "2", name: "Category 2" },
      ],
      sys: {
        cts: "2020-01-01T00:00:00.000Z",
      },
    },
    handlerMap: {
      update: jest.fn(),
    },
  };

  return { jokeDataObject, onSaveDone: jest.fn(), onCancel: jest.fn(), shown: true };
}

async function setup(props, options) {
  // ISSUE - uu5g05 - Uu5Forms.Form.View - invalid propTypes of Lsi
  // https://uuapp.plus4u.net/uu-sls-maing01/e80acdfaeb5d46748a04cfc7c10fdf4e/issueDetail?id=62f67d7f0b17bf002af36cfc
  omitConsoleLogs(
    "Warning: Failed prop type: Invalid prop `lsi` of type `string` supplied to `Uu5g05.Lsi`, expected `object`"
  );

  global.URL.createObjectURL = jest.fn();

  Client.get.mockReturnValueOnce({
    data: { itemList: categoryList, pageInfo: { pageSize: 10, pageIndex: 0, total: 3 } },
  }); // category/list

  Client.get.mockReturnValueOnce({
    itemList: categoryList,
    pageInfo: { pageSize: 10, pageIndex: 0, total: 3 },
  }); // category/list

  const customProps = { ...getDefaultProps(), ...props };
  const customOpts = { queries, ...options };
  const { view, ...setupResult } = await VisualComponent.setup(UpdateModal, customProps, customOpts);
  const submitBtn = screen.getByRole("button", { name: lsi.submit });
  const cancelBtn = screen.getByRole("button", { name: lsi.cancel });

  const inputs = {
    name: screen.getByLabelText(lsi.name),
    imageClearBtn: view.getFormsFileClearBtn(lsi.image),
    text: screen.getByLabelText(lsi.text),
    category: view.getFormsSelect(lsi.category),
    categoryClearBtn: view.getFormsSelectClearBtn(lsi.category),
  };

  return { ...setupResult, view, submitBtn, cancelBtn, inputs };
}

describe(`UuJokesCore.Joke.DetailView.UpdateModal`, () => {
  it("submits form without changes", async () => {
    const { user, props, submitBtn } = await setup();

    await user.click(submitBtn);

    expect(props.jokeDataObject.handlerMap.update).toHaveBeenCalledTimes(1);
    expect(props.jokeDataObject.handlerMap.update.mock.lastCall[0]).toMatchSnapshot();
    expect(props.onSaveDone).toHaveBeenCalledTimes(1);
  });

  it("submits form with changes", async () => {
    const { user, props, submitBtn, inputs } = await setup();

    await user.clear(inputs.name);
    await user.type(inputs.name, "Updated name");

    await user.click(inputs.imageClearBtn);

    await user.clear(inputs.text);
    await user.type(inputs.text, "Updated text");

    await user.click(inputs.categoryClearBtn);
    await user.click(inputs.category);

    const item = screen.getByRole("option", { name: "Category 3" });
    await user.click(item);

    await user.click(submitBtn);

    expect(props.jokeDataObject.handlerMap.update).toHaveBeenCalledTimes(1);
    expect(props.jokeDataObject.handlerMap.update.mock.lastCall[0]).toMatchSnapshot();
    expect(props.onSaveDone).toHaveBeenCalledTimes(1);
  });

  it("submits form with error", async () => {
    const { user, props, submitBtn } = await setup();

    UpdateModal.logger.error = jest.fn();
    props.jokeDataObject.handlerMap.update.mockImplementation(() => {
      throw new UuJokesError("baseNetworkError", "Test update error");
    });

    await user.click(submitBtn);

    expect(screen.getByText(enLsi.Errors.baseNetworkError)).toBeVisible();
    expect(UpdateModal.logger.error.mock.lastCall).toMatchSnapshot();
    expect(props.jokeDataObject.handlerMap.update).toHaveBeenCalledTimes(1);
    expect(props.onSaveDone).toHaveBeenCalledTimes(0);
  });

  it("stops submition due to invalid text and image", async () => {
    const { user, props, submitBtn, inputs } = await setup();

    await user.click(inputs.imageClearBtn);
    await user.clear(inputs.text);
    await user.click(submitBtn);

    expect(screen.getByText(lsi.textOrFile)).toBeVisible();
    expect(props.jokeDataObject.handlerMap.update).toHaveBeenCalledTimes(0);
    expect(props.onSaveDone).toHaveBeenCalledTimes(0);
  });

  it("cancels form", async () => {
    const { user, props, cancelBtn } = await setup();

    await user.click(cancelBtn);

    expect(props.jokeDataObject.handlerMap.update).toHaveBeenCalledTimes(0);
    expect(props.onCancel).toHaveBeenCalledTimes(1);
  });
});
