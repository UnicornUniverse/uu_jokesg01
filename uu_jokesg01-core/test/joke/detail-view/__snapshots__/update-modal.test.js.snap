// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UuJokesCore.Joke.DetailView.UpdateModal submits form with changes 1`] = `
Object {
  "categoryIdList": Array [
    "3",
  ],
  "deleteImage": true,
  "id": "123",
  "name": "Updated name",
  "text": "Updated text",
}
`;

exports[`UuJokesCore.Joke.DetailView.UpdateModal submits form with error 1`] = `
Array [
  "Error submitting form",
  [UuJokesError: Test update error],
]
`;

exports[`UuJokesCore.Joke.DetailView.UpdateModal submits form without changes 1`] = `
Object {
  "categoryIdList": Array [
    "1",
    "2",
  ],
  "id": "123",
  "image": File {},
  "name": "Test joke",
  "text": "Best joke ever",
}
`;
